firefox_ui_harness/testcase.py

```diff
@@ -5,23 +5,37 @@
 from marionette import MarionetteTestCase
 
 from firefox_puppeteer import Puppeteer
+from firefox_puppeteer.ui.windows import BrowserWindow
 
 
 class FirefoxTestCase(MarionetteTestCase, Puppeteer):
-    """
-    Test case that inherits from a Puppeteer object so Firefox specific
-    libraries are exposed to test scope.
+    """Base testcase class for Firefox Desktop tests.
+
+    It enhances the Marionette testcase by inserting the Puppeteer mixin class,
+    so Firefox specific API modules are exposed to test scope.
     """
     def __init__(self, *args, **kwargs):
         MarionetteTestCase.__init__(self, *args, **kwargs)
 
+    def _init_data(self):
+        self.marionette.set_context('chrome')
+
+        self.browser = self.windows.switch_to(lambda win: type(win) is BrowserWindow)
+
+    def restart(self, flags=None):
+        """Restart Firefox and re-initialize data.
+
+        :param flags: Specific restart flags for Firefox
+        """
+        self.marionette.restart()
+        self._init_data()
```

Comment:
> Ok, so can we move this helper inline here and not add the additional
> check to setUp? If a test is switched into the wrong window at setUp it
> means the previous test has leaked a window and we're unlikely to
> succeed by virtue of switching back to a browser window.

What's the problem for at least guarding our harness to continue with
tests if at least one browser window is open, but which is not the
current window. If we have code like you proposed, we would mark all the
following tests as broken.

In tearDown we throw in case of not all additional windows have been
closed. I think that is a wonderful indicator if a test needs a fix. But
shall all remaining ones really suffer from that breakage if we have a
chance to pick up a browser window?